apps/aecore/src/aec_fork_signalling.erl

```diff
@@ -0,0 +1,359 @@
+-module(aec_fork_signalling).
+
+-behaviour(gen_server).
+
+%% API
+-export([start_link/0,
+         get_fork_result/2,
+         compute_fork_result/2
+        ]).
+
+%% gen_server callbacks
+-export([init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2
+        ]).
+
+-export([worker_process/4]).
+
+-define(SERVER, ?MODULE).
+
+-type header_hash()      :: aec_blocks:block_header_hash().
+
+-type worker()           :: {pid(), reference()}.
+
+-type result()           :: true | false | pending.
+
+%% The pointer is a key block hash which belongs to the last signalling block
+%% (the block at height HE - 1).
+-type pointer()          :: header_hash().
+
+-record(state, {fork     :: aec_hard_forks:fork() | undefined,
+                results  :: #{header_hash() := result() | pointer()},
+                workers  :: #{worker() := header_hash()}
+               }).
+
+-define(IS_RESULT(X), X =:= true orelse X =:= false orelse X =:= pending).
+-define(IS_POINTER(X), is_binary(X)).
+
+%% API
+
+start_link() ->
+    gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
+
+get_fork_result(Block, BlockHash) ->
+    gen_server:call(?SERVER, {get_fork_result, Block, BlockHash}).
+
+compute_fork_result(Block, BlockHash) ->
+    gen_server:cast(?SERVER, {compute_fork_result, Block, BlockHash}).
+
+%% gen_server callbacks
+
+init([]) ->
+    process_flag(trap_exit, true),
+    {ok, #state{fork = aec_hard_forks:fork(), results = maps:new(),
+                workers = maps:new()}}.
+
+handle_call({get_fork_result, Block, BlockHash}, _From, State) ->
+    {Reply, State1} = handle_get_fork_result(Block, BlockHash, State),
+    {reply, Reply, State1};
+handle_call(_Request, _From, State) ->
+    {reply, ingnored, State}.
+
+handle_cast({compute_fork_result, Block, BlockHash}, State) ->
+    {noreply, handle_compute_fork_result(Block, BlockHash, State)};
+handle_cast(_Request, State) ->
+    {noreply, State}.
+
+handle_info({worker_msg, Pid, Msg}, State) ->
+    {noreply, handle_worker_msg(Pid, Msg, State)};
+handle_info({'DOWN', Ref, process, Pid, Rsn}, State) ->
+    {noreply, handle_down({Pid, Ref}, Rsn, State)};
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+terminate(_Rsn, #state{workers = Workers} = State) ->
+    maps:foldl(fun(W, _BH, S1) -> kill_worker(W, S1) end, State, Workers).
+
+%% Internal functions
+
+handle_get_fork_result(Block, BlockHash,
+                       #state{fork = Fork, results = Results} = State) ->
+    case is_last_block_before_fork(Block, Fork) of
+        true ->
+            case get_result(BlockHash, Results) of
+                R when ?IS_RESULT(R) ->
+                    %% The block at height H2 - 1 has a result (not a pointer).
+                    %% It means that the block at height HE - 1 is the same as
+                    %% H2 - 1. There are no blocks between, so HE < H2 doesn't
```

Comment:
Shall this clause ever apply then?

I would then expect this code to either:
- Rely on constraint having been enforced by user configuration load, and not have this clause at all.
- Return an error that clearly signals misconfiguration, and then caller code shall log error (blocking, though shall never happen).

Reply:
Fixed in 88ca0f4c3.